File: go-backend-interview-prep/04_interfaces_and_generics/04_type_switch/main.py
# Python equivalent of Type Switch (compare with main.go)
# Go type switch ~ Python match/case with type patterns (3.10+) or isinstance chains


def classify(v: object) -> None:
    """Python 3.10+ structural pattern matching on types."""
    match v:
        case str():
            print(f"  string: {v!r} (len={len(v)})")
        case bool():
            # Must check bool before int (bool is subclass of int)
            print(f"  bool: {v}")
        case int():
            print(f"  int: {v}")
        case float():
            print(f"  float: {v:.2f}")
        case None:
            print("  None")
        case _:
            print(f"  unknown: {v} ({type(v).__name__})")


def is_numeric(v: object) -> bool:
    """Multiple types in one check."""
    return isinstance(v, (int, float)) and not isinstance(v, bool)


def main():
    # --- Basic type dispatch ---
    print("--- Type switch ---")
    values: list[object] = [42, "hello", 3.14, True, None, [1, 2]]
    for v in values:
        classify(v)

    # --- Multiple types check ---
    print("\n--- is_numeric ---")
    print("42:", is_numeric(42))
    print("3.14:", is_numeric(3.14))
    print("hello:", is_numeric("hello"))

    # --- isinstance chains (pre-3.10 style) ---
    print("\n--- isinstance chains ---")
    v: object = "test"
    if isinstance(v, str):
        print(f"  str: {v}")
    elif isinstance(v, int):
        print(f"  int: {v}")
    else:
        print(f"  other: {v}")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/04_interfaces_and_generics/04_type_switch/main.go
package main

import "fmt"

func classify(v any) {
	switch val := v.(type) {
	case string:
		fmt.Printf("  string: %q (len=%d)\n", val, len(val))
	case int:
		fmt.Printf("  int: %d\n", val)
	case float64:
		fmt.Printf("  float64: %.2f\n", val)
	case bool:
		fmt.Printf("  bool: %t\n", val)
	case nil:
		fmt.Println("  nil")
	default:
		fmt.Printf("  unknown: %v (%T)\n", val, val)
	}
}

// --- Multiple types in one case ---
func isNumeric(v any) bool {
	switch v.(type) {
	case int, int8, int16, int32, int64,
		float32, float64:
		return true
	default:
		return false
	}
}

// --- Interface type switch ---
type Stringer interface {
	String() string
}

type Named struct{ Name string }

func (n Named) String() string { return n.Name }

func describeInterface(v any) {
	switch val := v.(type) {
	case Stringer:
		fmt.Printf("  implements Stringer: %s\n", val.String())
	case error:
		fmt.Printf("  implements error: %s\n", val.Error())
	default:
		fmt.Printf("  no known interface: %T\n", val)
	}
}

func main() {
	// --- Basic type switch ---
	fmt.Println("--- Type switch ---")
	values := []any{42, "hello", 3.14, true, nil, []int{1, 2}}
	for _, v := range values {
		classify(v)
	}

	// --- Multiple types in one case ---
	fmt.Println("\n--- isNumeric ---")
	fmt.Println("42:", isNumeric(42))
	fmt.Println("3.14:", isNumeric(3.14))
	fmt.Println("hello:", isNumeric("hello"))

	// --- Interface type switch ---
	fmt.Println("\n--- Interface type switch ---")
	describeInterface(Named{Name: "Alice"})
	describeInterface(fmt.Errorf("something failed"))
	describeInterface(42)
}
